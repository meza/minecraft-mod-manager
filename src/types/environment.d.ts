declare global {
  // eslint-disable-next-line no-unused-vars
  namespace NodeJS {
    // eslint-disable-next-line no-unused-vars
    interface ProcessEnv {
      CURSEFORGE_API_KEY: string,
      MODRINTH_API_KEY: string,
      POSTHOG_API_KEY: string,
      npm_package_version: string
    }
  }
}

export {};
