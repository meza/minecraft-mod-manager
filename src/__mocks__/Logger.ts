import { vi } from 'vitest';
import { Logger as Original } from '../lib/Logger.js';

export const Logger = vi.importMock('../lib/Logger.js') as unknown as Original;
