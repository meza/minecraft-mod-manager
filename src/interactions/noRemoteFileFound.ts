import { confirm, input } from '@inquirer/prompts';
import chalk from 'chalk';
import { Logger } from '../lib/Logger.js';
import { ModsJson, Platform } from '../lib/modlist.types.js';
import { DefaultOptions } from '../mmm.js';

interface NoRemoteFileFoundInteractionResult {
  platform: Platform;
  id: string;
}

export const noRemoteFileFound = async (
  modName: string,
  platform: Platform,
  configuration: ModsJson,
  logger: Logger,
  options: DefaultOptions
): Promise<NoRemoteFileFoundInteractionResult> => {
  const errorText =
    `Could not find a file for ${modName} and the Minecraft version ${chalk.whiteBright(configuration.gameVersion)} ` +
    `for ${chalk.whiteBright(configuration.loader)} loader`;

  if (options.quiet === true) {
    logger.error(errorText);
  }

  const newPlatform = platform === Platform.CURSEFORGE ? Platform.MODRINTH : Platform.CURSEFORGE;

  const answer = await confirm({
    message: `${errorText}. Would you like to try on ${newPlatform}?`,
    default: true
  });

  if (answer === false) {
    logger.error('Aborting', 0);
  }

  const newModName = await input({
    message: `What is the project id of the mod you want to add on ${newPlatform}?`
  });

  return {
    id: newModName,
    platform: newPlatform
  };
};
