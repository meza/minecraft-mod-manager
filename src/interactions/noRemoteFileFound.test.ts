import { confirm, input } from '@inquirer/prompts';
import { chance } from 'jest-chance';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { generateRandomPlatform } from '../../test/generateRandomPlatform.js';
import { generateModsJson } from '../../test/modlistGenerator.js';
import { Logger } from '../lib/Logger.js';
import { Loader, Platform } from '../lib/modlist.types.js';
import { noRemoteFileFound } from './noRemoteFileFound.js';

vi.mock('@inquirer/prompts');
vi.mock('../mmm.js');
vi.mock('../lib/Logger.js');

describe('The mod not found interaction', () => {
  let logger: Logger;

  beforeEach(() => {
    logger = new Logger({} as never);
    vi.mocked(logger.error).mockImplementation(() => {
      throw new Error('process.exit');
    });
  });
  afterEach(() => {
    vi.resetAllMocks();
  });

  it('shows the correct error message when in quiet mode', async () => {
    const testPlatform = Platform.CURSEFORGE;
    const testModId = 'test-mod-id';
    const randomConfig = generateModsJson({
      gameVersion: '1.16.5',
      loader: Loader.FORGE
    }).generated;

    await expect(
      noRemoteFileFound(testModId, testPlatform, randomConfig, logger, {
        config: 'config.json',
        quiet: true
      })
    ).rejects.toThrow(new Error('process.exit'));

    const loggerErrorCall = vi.mocked(logger.error).mock.calls[0][0];

    expect(loggerErrorCall).toMatchInlineSnapshot(
      '"Could not find a file for test-mod-id and the Minecraft version 1.16.5 for forge loader"'
    );
    expect(vi.mocked(confirm)).not.toHaveBeenCalled();
    expect(vi.mocked(input)).not.toHaveBeenCalled();
  });

  it('aborts when the user does not want to modify their search', async () => {
    const testPlatform = generateRandomPlatform();
    const testModId = chance.word();
    const randomConfig = generateModsJson().generated;

    vi.mocked(confirm).mockResolvedValueOnce(false);

    await expect(
      noRemoteFileFound(testModId, testPlatform, randomConfig, logger, {
        config: 'config.json',
        quiet: false
      })
    ).rejects.toThrow(new Error('process.exit'));

    const loggerErrorCall = vi.mocked(logger.error).mock.calls[0][0];

    expect(loggerErrorCall).toMatchInlineSnapshot('"Aborting"');
    expect(vi.mocked(confirm)).toHaveBeenCalledTimes(1);
    expect(vi.mocked(input)).not.toHaveBeenCalled();
  });

  describe.each([
    { userInput: Platform.CURSEFORGE, expected: Platform.MODRINTH },
    { userInput: Platform.MODRINTH, expected: Platform.CURSEFORGE }
  ])('when the user wants to modify their search for $input', ({ userInput, expected }) => {
    it(`it asks for ${expected} when they come from ${input}`, async () => {
      const testPlatform = userInput;
      const testModId = chance.word();
      const randomConfig = generateModsJson().generated;

      vi.mocked(confirm).mockResolvedValueOnce(true);
      vi.mocked(input).mockResolvedValueOnce('new-mod-id');

      const actual = await noRemoteFileFound(testModId, testPlatform, randomConfig, logger, {
        config: 'config.json',
        quiet: false
      });

      const inquirerPromptCallArgs = vi.mocked(input).mock.calls[0][0];

      expect(actual).toEqual({ id: 'new-mod-id', platform: expected });

      expect(inquirerPromptCallArgs).toMatchSnapshot();
    });
  });
});
