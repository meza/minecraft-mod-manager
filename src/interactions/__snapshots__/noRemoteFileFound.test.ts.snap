// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The mod not found interaction > when the user wants to modify their search for 'curseforge' > it asks for modrinth when they come from curseforge 1`] = `
[
  {
    "message": "What is the project id of the mod you want to add on modrinth?",
    "name": "newModName",
    "type": "input",
  },
]
`;

exports[`The mod not found interaction > when the user wants to modify their search for 'modrinth' > it asks for curseforge when they come from modrinth 1`] = `
[
  {
    "message": "What is the project id of the mod you want to add on curseforge?",
    "name": "newModName",
    "type": "input",
  },
]
`;
